src/couch/src/couch_doc.erl

```diff
@@ -136,12 +136,27 @@ from_json_obj_validate(EJson, DbName) ->
     case couch_ejson_size:encoded_size(Doc#doc.body) =< MaxSize of
         true ->
              validate_attachment_sizes(Doc#doc.atts),
+             validate_total_document_size(Doc),
              Doc;
         false ->
             throw({request_entity_too_large, Doc#doc.id})
     end.
 
 
+% sum up the json body size + attachment body size and
+% make sure it is < max_http_request_size
+validate_total_document_size(#doc{id=DocId, body=Body, atts=Atts0}) ->
+    MaxReqSize = config:get_integer("httpd", "max_http_request_size", 4294967296), % 2 GB
+    Boundary = couch_uuids:random(), % mock boundary, is only used for the length
+    Atts = lists:map(fun couch_att:to_tuple/1, Atts0),
+    {_, DocSum} = couch_httpd_multipart:length_multipart_stream(Boundary,
+        ?JSON_ENCODE(Body), Atts),
```

Comment:
When we re-encode, we use jiffy, but unless it is a request made by the replicator, the user probably didn't use erlang to encode the data, so we could get a different value. There is also some performance loss in say re-encoding larger document bodies back to json just to check their size.

There is no canonical json encoding and so no canonical encoded json size. Above we are calculating the encoded size using a conservative size estimate (giving the user a benefit of the doubt) and with better performance (https://github.com/nickva/couch_ebench).  Maybe make a version of length calculation that takes sizes and then we'd pass the already computed `couch_ejson_size:encoded_size(Doc#doc.body)` and 32 for boundary size.

Reply:
Thanks @nickva, I’ll give this a ponder!